### PR TITLE
Add `Controller::for_stream` + `Controller::watches_stream`

### PR DESCRIPTION
**Pull request by clux**
This adds a stream interface analogue for `Controller::new`, that allows the user to create the root `watcher` and `reflector`, outside, and pass in the reader + flattened stream. This is for #1180 as a building block for stream sharing #1080.

In a similar vein, we also add `Controller::watches_stream` as an analogue for `Controller::watches`. That name should be less controversial given we have `Controller::owns_stream`.

### Naming
For now, have named it along the lines of `Controller::for` and highlighted the terminology (noting that `Controller::For` is the [controller-runtime equivalent](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/builder#Builder.For)).

However, there may be some better alternatives. Previously suggested:

- `Controller::from_stream` - original plan, but that was before we needed to pass in the `reader` - so not a clean `From`
- `Controller::new_stream` - awkward, suggests we are creating a new stream?
- `Controller::new_from_stream` - very long, and leads to the obscene `::new_from_stream_with`
- `Controller::new_with` - taken by the dynamic variant
- `Controller::new_via` - also awkward in combination with `::new_via_with`

Feel like realistically it's between `Controller::for_stream` and `Controller::from_stream`.

### Docs Fixes + Unstable Feature Consolidation

Also fixes up bad doc setups and ensures unstable features are highlighted, even though they appear in the middle of a trait (i am not convinced it will show up correctly on docs.rs so being a little careful first time).

The owns-stream unstable flagged is moved to `unstable-runtime-stream-share`, which now also contains this new method.

### Consequences
When using `unstable-runtime`, we can actually use `metadata_watcher` with all `Controller` input interfaces.
Tested on a controller running several controller loops on a large cluster with thousands of pods and saw a >2x memory drop (449MB -> 210MB from a smaller core controller reflector passed in), and significant bandwidth reduction as well (230kB/s -> 30kB/s TX, 20kB/s -> 13kB/s RX but that's partially from using predicates).

### Stream Sharing
Currently this interface does not work with `WatchStreamExt::stream_subscribe` because of the lack of `Arc<K>` in our interfaces. I want to change this long term, but this also involves adding `Arc` on everything internal to `Controller`, but I think that's OK. These new interfaces are unstable, and streams are otherwise constructed internally and could be arc'd up therein.

We already `Arc`-wrap before calling out to `reconcile` and `error_policy` so it would be nice to just have `watcher` (or a method in `WatchStreamExt` do this. Leaving this for a separate PR though.

### CODE REVIEW

**Comment by clux** [comment]
This is working good enough for an unstable inclusion imo, but can benefit from a look-over.

It will not work for stream sharing yet because we need consensus on whether to Arc up literally every stream interface, but this at the very least allows configuring streams outside `Controller` and use filters or metadata_watcher with controllers.

Pinging @Dav1dde @danrspencer as interested parties.

**Comment by Dav1dde** [comment]
I think `for_stream` is good, since it requires multiple arguments, it's also not a 'clean' from anymore.

File: kube-runtime/src/controller/mod.rs [hunk 1]
```diff
@@ -755,6 +848,98 @@ where
...
+    ) -> Self
+    where
+        Other: Clone + Resource + DeserializeOwned + Debug + Send + 'static,
+        Other::DynamicType: Debug + Clone,
+        I: 'static + IntoIterator<Item = ObjectRef<K>>,
+        I::IntoIter: Send,
+    {
+        let other_watcher = trigger_with(trigger, move |obj| {
+            let watched_obj_ref = ObjectRef::from_obj_with(&obj, dyntype.clone()).erase();
+            mapper(obj)
```
Comment (Dav1dde) [inline kube-runtime/src/controller/mod.rs:930]: This is very similar to `trigger_owners`, can't we implement `trigger_owners` using this by just supplying a mapper?

Maybe move this out into a standalone `trigger_*` function and implement `trigger_owners` using it?

Comment (clux) [reply kube-runtime/src/controller/mod.rs]: Yeah, you're right. The pattern is that most of the bodies of the `Controller::methods` are factored out into trigger helpers in:
https://github.com/kube-rs/kube/blob/36dcdaaad178d4f6cbe0bd7513d9608bfcf09bb6/kube-runtime/src/controller/mod.rs#L87-L150

ended up doing https://github.com/kube-rs/kube/pull/1187/commits/fc77780f8bbcebb32633f8a1e2c8fe4de4749f3a which has a fairly horrendous generic signature, but it's nice to have it out of the imperative `Controller` logic.

I think long term all these trigger helpers can be turned into `WatchStreamExt` methods as that will make it chain nicer. Will open an issue for this.

**Comment by Dav1dde** [comment]
`trigger_owners` can be implemented as this now, I think:

```rs
/// Enqueues any owners of type `KOwner` for reconciliation
pub fn trigger_owners<KOwner, S>(
    stream: S,
    owner_type: KOwner::DynamicType,
    child_type: <S::Ok as Resource>::DynamicType,
) -> impl Stream<Item = Result<ReconcileRequest<KOwner>, S::Error>>
where
    S: TryStream,
    S::Ok: Resource,
    <S::Ok as Resource>::DynamicType: Clone,
    KOwner: Resource,
    KOwner::DynamicType: Clone,
{
    let mapper = move |obj: S::Ok| {
        let meta = obj.meta().clone();
        let ns = meta.namespace;
        let owner_type = owner_type.clone();
        meta.owner_references
            .into_iter()
            .flatten()
            .filter_map(move |owner| ObjectRef::from_owner_ref(ns.as_deref(), &owner, owner_type.clone()))
    };
    trigger_others(stream, mapper, child_type)
}
```

**Comment by clux** [comment]
That's great. Worked immediately.

**Comment by clux** [comment]
Alright windows is green here after sync despite it consistently crashing in the other PR :upside_down_face:

Will merge this at lunch and prepare a release.

**Pull request merged**